README.txt
Copyright (C) 2015 Alex Henry

This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License version 2 as published by the Free Software Foundation.

This program uses rules and monster from the Wizard's of the Coast d20 System Reference Document, released under the Open Gaming License but it does not offer open content of its own.

Full copies of both licenses are available under the ‘doc’ directory.

For more information and to keep updated visit https://javelinrl.wordpress.com/


doc/CHANGELOG.txt
Version 1.7

 * Mouse support!
 * Game sounds!
 * 3 gameplay modes: Campaign, Dungeon World and Art of War!
 * 2 mini-games: Arena and Battlefield!
 * City management redesign, with 7 city traits and 36 projects!
 * New systems for town quests, diplomacy, urban and wilderness events!
 * New items: caster rings (grants spells) and Eidolons (summons allies)!
 * Dungeon redesign (templated-based maps, 4 tiers, keys, NPCs and more)!
 * Introducing 6 Haunt locations!
 * New kits: 8 wizards (one per school of magic), 5 dragoons and more!
 * Items and Upgrades no longer tied to Realms (upgrades now fully kit-based)
 * Hostile monsters and mercenaries can now be Elite or kit-based NPCs
 * Martial disciplines (Steel Serpent)
 * Hostile towns now spawn incursions that can capture locations (campaign)
 * New locations: natural resources, Assassins Guild and Deep Dungeon
 * New skills: Bluff and Sense Motive
 * New world action: Infiltrate hostile locations
 * World map fog-of-war
 * More responsive user-interface during battles
 * Improved battle AI
 * Multi-threaded world generation
 * Auto-attack using the Tab key (for both close and ranged combat)
 * Improved artwork
 * Gems and Art Objects as treasure
 * Downloads now come with self-contained, optmized per-OS Java 12 runtimes

Version 1.6 (2016-07-15) 

 * New locations: 7 temples, 11 mage guilds, troves and mines!
 * New starting locations: Training Hall and Adventurers Guild!
 * New unique locations: Summoning Circle and Pillar of Skulls!
 * Around 50 new monsters!
 * New end-game goal: bring the 7 Relics to Haxor!
 * All game images can now be easily modified! (mod/skin support)
 * New damage effects: poison, paralysis, fear
 * Burrow movement
 * New item type: wands
 * New skills: use magic device and heal
 * New transport: sailing ship
 * More realistic, dynamic land movement
 * Seasons, new terrains and terrain features
 * Town workers let you build roads, settlements and more
 * Player units can now be set to automatic (controlled by AI)
 * New party option: strategic combat (skips random encounters)
 * Preferences and command keys can now be edited in-game
 * In-game journal for taking notes
 * More information on errors, for easier bug reporting
 * Improved save system, also creates automatic backups
 * AI system now monitors thinking speed and corrects itself if slow
 * 183 monsters
 * 47 items, 12 artifacts
 * 63 upgrades, 49 spells, 13 skills
 * 22 battle maps

Version 1.5 (2016-05-06)

 * Bug-fix release

Version 1.4 (2016-04-30)

 * Bug-fix release

Version 1.3 (2016-04-29)

 * 6 independent, opposing factions to play against at every game!
 * City-building, automatic or manual!
 * Equippable artifacts!
 * 6 new location types to explore on the world map!
 * 2 new unique locations: mercenary guild and artificer!
 * Skill system, featuring 13 different skills!
 * New units can now be acquired by spending XP
 * The world map now has to be explored
 * Dungeon redesign, featuring traps
 * 137 monsters
 * 28 items, 33 artifacts
 * 53 upgrades, 25 spells, 13 skills

Version 1.2 (2016-02-29)

 * Minor release in response to community feedback on 1.1
 * Improved documentation
 * Breath area now ignores cover
 * AI cache is now disabled by default (can slow down some systems)

Version 1.1 (2016-02-26)

 * Dungeons!
 * Vehicles!
 * Portals!
 * Tournaments!
 * The Temple of Haxor!
 * Many new, randomly-generated battle maps!
 * Lots of UI and AI improvements!
 * Each town now has a unique selection of items and upgrades
 * Towns now take a certain amount of time to craft items and train heroes
 * In-battle power-ups for special battles
 * A total of 136 monsters, 27 items, 57 upgrades and 25 spells
 
1.0 (2015-09-01)

 * Javelin launched!


javelin/Javelin.java
package javelin;

import java.awt.BorderLayout;
import java.awt.Frame;
import java.awt.Image;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;
import java.io.FileReader;
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Set;

import javax.swing.JFrame;
import javax.swing.WindowConstants;
import javax.xml.parsers.SAXParserFactory;

import org.xml.sax.InputSource;
import org.xml.sax.XMLReader;

import javelin.controller.CountingSet;
import javelin.controller.Highscore;
import javelin.controller.db.Preferences;
import javelin.controller.db.StateManager;
import javelin.controller.db.reader.MonsterReader;
import javelin.controller.db.reader.fields.Organization;
import javelin.controller.upgrade.classes.ClassLevelUpgrade;
import javelin.model.item.Item;
import javelin.model.item.artifact.Artifact;
import javelin.model.unit.Combatant;
import javelin.model.unit.Squad;
import javelin.model.unit.abilities.spell.Spell;
import javelin.model.unit.abilities.spell.conjuration.Summon;
import javelin.model.world.Actor;
import javelin.model.world.World;
import javelin.model.world.location.Academy;
import javelin.old.Interface;
import javelin.old.messagepanel.MessagePanel;
import javelin.old.messagepanel.TextZone;
import javelin.view.Images;
import javelin.view.LauncherDialog;
import javelin.view.screen.BattleScreen;
import javelin.view.screen.InfoScreen;
import javelin.view.screen.Option;
import javelin.view.screen.WorldScreen;
import javelin.view.screen.town.SelectScreen;

/**
 * Utility class for broad-level rules and game-behavior. Add the VM argument
 * -Ddebug=true to the java command line for easier debugging and logging.
 *
 * @see #DEBUG
 * @author alex
 */
public class Javelin{
	public enum Delay{
		NONE,WAIT,BLOCK
	}

	/**
	 * Add -Ddebug=true to the java VM command line for easier debugging and
	 * logging.
	 */
	public static final boolean DEBUG=System.getProperty("debug")!=null;

	/** TODO turn into {@link Enum} */
	public static final String PERIODMORNING="Morning";
	/** TODO turn into {@link Enum} */
	public static final String PERIODNOON="Noon";
	/** TODO turn into {@link Enum} */
	public static final String PERIODEVENING="Evening";
	/** TODO turn into {@link Enum} */
	public static final String PERIODNIGHT="Night";
	public static final String[] PERIODS=new String[]{PERIODMORNING,PERIODNOON,
			PERIODEVENING,PERIODNIGHT};

	public static final Image[] ICONS=new Image[]{Images.get("javelin")};

	/** Roll target on a d20. */
	public static final int HARD=16;
	/** Roll target on a d20. */
	public static final int FAIR=12;
	/** Roll target on a d20. */
	public static final int EASY=8;
	/** Roll target on a d20. */
	public static final int EFFORTLESS=4;

	static final String TITLE="Javelin";
	static final DecimalFormat COSTFORMAT=new DecimalFormat("####,###,##0");

	/** Singleton. */
	public static JavelinApp app;
	public static boolean delayblock=false;

	static{
		ClassLevelUpgrade.setup();
		Spell.setup();
		try{
			final MonsterReader monsterdb=new MonsterReader();
			final XMLReader xml=SAXParserFactory.newInstance().newSAXParser()
					.getXMLReader();
			xml.setContentHandler(monsterdb);
			xml.setErrorHandler(monsterdb);
			FileReader filereader=new FileReader("monsters.xml");
			xml.parse(new InputSource(filereader));
			filereader.close();
		}catch(final Exception e){
			throw new RuntimeException(e);
		}
		Organization.setup();
		MonsterReader.closelogs();
		Summon.setupsummons();
		Spell.setup();
		Artifact.setup();
		Item.setup();
	}

	/**
	 * First method to be called.
	 *
	 * @param args See {@link #DEBUG}.
	 */
	public static void main(final String[] args){
		Thread.currentThread().setName("Javelin");
		app=new JavelinApp();
		LauncherDialog.choose(args);
		final JFrame f=new JFrame(TITLE);
		f.setBackground(java.awt.Color.black);
		f.addWindowListener(StateManager.SAVEONCLOSE);
		f.setDefaultCloseOperation(WindowConstants.DO_NOTHING_ON_CLOSE);
		f.setLayout(new BorderLayout());
		f.setFocusTraversalKeysEnabled(false);
		f.setIconImages(Arrays.asList(ICONS));
		app.frame=f;
		app.setVisible(false);
		f.add(app);
		f.setSize(app.getPreferredSize().width,app.getPreferredSize().height);
		f.addKeyListener(new KeyAdapter(){
			@Override
			public void keyPressed(final KeyEvent e){
				Interface.userinterface.go(e);
			}
		});
		app.setVisible(true);
		f.setVisible(true);
		f.setExtendedState(Frame.MAXIMIZED_BOTH);
		app.init();
	}

	/**
	 * TODO move to a new class with proper enum?
	 *
	 * @return {@link #PERIODEVENING}, {@link #PERIODMORNING},
	 *         {@link #PERIODNIGHT} or {@value #PERIODNOON}.
	 */
	public static String getperiod(){
		if(Javelin.app.fight!=null) return Javelin.app.fight.period;
		final long hourofday=getHour();
		if(hourofday<6) return PERIODNIGHT;
		if(hourofday<12) return PERIODMORNING;
		if(hourofday<18) return PERIODNOON;
		return PERIODEVENING;
	}

	/**
	 * If {@link Squad#active} is <code>null</code> for any reason, will return
	 * zero.
	 *
	 * @return Hour of the day, from 0 to 23.
	 */
	public static long getHour(){
		return Squad.active==null?0:Squad.active.gettime()%24;
	}

	/**
	 * This is also the only function that should write to {@link Squad#active} so
	 * it should be called with extreme caution to avoid changing it in the middle
	 * of an action.
	 *
	 * @return Next squad to act.
	 */
	public static Squad act(){
		Squad next=nexttoact();
		Squad.active=next;
		if(WorldScreen.lastday==-1)
			WorldScreen.lastday=Math.ceil(Squad.active.gettime()/24.0);
		return next;
	}

	/**
	 * @return Next squad to act.
	 * @see Squad#time
	 */
	public static Squad nexttoact(){
		Squad next=null;
		for(final Actor a:World.getall(Squad.class)){
			Squad s=(Squad)a;
			if(next==null||s.gettime()<next.gettime()) next=s;
		}
		return next;
	}

	/**
	 * Pure fluff/flavor.
	 *
	 * @return Welcomes the playet to the game based on the current time of the
	 *         day.
	 */
	public static String welcome(){
		final String period=getperiod();
		String flavor;
		if(period==PERIODMORNING)
			flavor="What dangers lie ahead..?";
		else if(period==PERIODNOON)
			flavor="Onwards to victory!";
		else if(period==PERIODEVENING)
			flavor="Cheers!";
		else if(period==PERIODNIGHT)
			flavor="What a horrible night to suffer an invasion...";
		else
			throw new RuntimeException("No welcome message");
		return "Welcome! "+flavor
				+"\n\n(press h at the overworld or battle screens for help)";
	}

	/**
	 * Once the player has no more {@link Squad}s and {@link Combatant}s under his
	 * control call this to stop the current game, invalidate the save file,
	 * record the highscore and exit the application.
	 */
	public static void lose(){
		if(Academy.train()) return;
		Javelin.app.switchScreen(BattleScreen.active);
		StateManager.clear();
		BattleScreen.active.messagepanel.clear();
		String sadface="You have lost all your units! Game over T_T\n\n";
		Javelin.message(sadface+Highscore.record(),Javelin.Delay.NONE);
		while(InfoScreen.feedback()!='\n')
			continue;
		System.exit(0);
	}

	/**
	 * @param rolltohit The number that needs to be rolled on a d20 for this
	 *          action to succeed.
	 * @return A textual representation of how easy or hard this action is to
	 *         achieve.
	 */
	static public String translatetochance(int rolltohit){
		if(rolltohit<=EFFORTLESS) return "effortless";
		if(rolltohit<=EASY) return "easy";
		if(rolltohit<=FAIR) return "fair";
		if(rolltohit<=HARD) return "hard";
		return "unlikely";
	}

	/**
	 * Utility function for user-input selection.
	 *
	 * TODO due to the UI being poor, if this method runs out of alphanumeric
	 * characters to use, it will display more choices but not allow the player to
	 * actually select them.
	 *
	 * @param prompt Text to show the user.
	 * @param names Will show each's {@link Object#toString()} as an option.
	 * @param fullscreen <code>true</code> to open in a new screen. Otherwise uses
	 *          the message panel.
	 * @param forceselection If <code>false</code> will allow the user to abort
	 *          the operation.
	 * @return The index of the selected element or -1 if aborted. Won't return an
	 *         invalid index other than -1.
	 */
	static public int choose(String prompt,List<?> names,boolean fullscreen,
			boolean forceselection){
		if(!forceselection) prompt+=" (q to quit)";
		prompt+="\n\n";
		int nnames=names.size();
		boolean multicolumn=nnames>20;
		ArrayList<Object> options=new ArrayList<>();
		for(int i=0;i<nnames;i++){
			boolean leftcolumn=i%2==0;
			String name=names.get(i).toString();
			options.add(name);
			String item="["+SelectScreen.getkey(i)+"] "+name;
			if(multicolumn&&leftcolumn) while(item.length()<50)
				item+=" ";
			prompt+=item;
			if(!multicolumn||!leftcolumn) prompt+="\n";
		}
		if(fullscreen)
			app.switchScreen(new InfoScreen(prompt));
		else{
			app.switchScreen(BattleScreen.active);
			MessagePanel.active.clear();
			Javelin.message(prompt,Javelin.Delay.NONE);
		}
		while(true)
			try{
				Character c=InfoScreen.feedback();
				if(!forceselection&&(c=='q'||c==InfoScreen.ESCAPE)) return -1;
				int selected=SelectScreen.convertkeytoindex(c);
				if(0<=selected&&selected<names.size()) return selected;
			}catch(Exception e){
				continue;
			}
	}

	/**
	 * Main output function for {@link WorldScreen}. Waits for user input for
	 * confirmation.
	 *
	 * @param text Prints this message in the status panel.
	 * @param requireenter If <code>true</code> will wait for the player to press
	 *          ENTER, otherwise any key will do.
	 * @return the key pressed by the user as confirmation for seeing the message.
	 */
	public static KeyEvent message(String text,boolean requireenter){
		var screen=BattleScreen.active==null?WorldScreen.current
				:BattleScreen.active;
		app.switchScreen(screen);
		//		screen.center();
		MessagePanel.active.clear();
		text+="\nPress "+(requireenter?"ENTER":"any key")+" to continue...";
		Javelin.message(text,Javelin.Delay.NONE);
		KeyEvent input=Javelin.input();
		while(requireenter&&input.getKeyChar()!='\n')
			input=Javelin.input();
		MessagePanel.active.clear();
		return input;
	}

	/**
	 * Prompts a message in the {@link WorldScreen}.
	 *
	 * @param prompt Text to show.
	 * @return Any {@link InfoScreen#feedback()}.
	 */
	static public Character prompt(final String prompt,boolean center){
		MessagePanel.active.clear();
		BattleScreen.active.center();
		Javelin.message(prompt,Javelin.Delay.NONE);
		if(center) BattleScreen.active.center();
		return InfoScreen.feedback();
	}

	/**
	 * @param prompt Shows this message...
	 * @return and returns the user input.
	 */
	static public Character prompt(final String prompt){
		return prompt(prompt,false);
	}

	public static String describedifficulty(int dc){
		if(dc<=0) return "very easy";
		if(dc<=5) return "easy";
		if(dc<=10) return "average";
		if(dc<=15) return "tough";
		if(dc<=20) return "challenging";
		if(dc<=25) return "formidable";
		if(dc<=30) return "heroic";
		return "nearly impossible";
	}

	/**
	 * @param message Shows this in a fullscreen, requires enter to leave.
	 */
	public static void show(String message){
		InfoScreen s=new InfoScreen("");
		s.print(message);
		while(s.getInput()!='\n'){
			// wait for enter
		}
	}

	public static char promptscreen(String prompt){
		InfoScreen s=new InfoScreen("");
		s.print(prompt);
		return InfoScreen.feedback();
	}

	/**
	 * @return Textual representation of the givne {@link Option#price}.
	 */
	static public String format(double gold){
		return COSTFORMAT.format(gold);
	}

	/**
	 * Example: 194,151 will be "rounded" to 190,000.
	 *
	 * A 10% loss of precision is expected at most but it tends to normalize (and
	 * lessen with the game's exponential increase as challenge levels rise). The
	 * loss of precision is easily worth the much higher readability.
	 *
	 * @param gold A value in gold or other unit that needn't be precise.
	 *
	 * @return The input value, but rounded off as to be more legible.
	 */
	static public int round(int gold){
		if(gold<=100) return gold;
		int roundto=1;
		while(roundto*100<gold)
			roundto=roundto*10;
		return roundto*Math.round((float)gold/roundto);
	}

	public static void redraw(){
		BattleScreen.active.mappanel.refresh();
		MessagePanel.active.repaint();
	}

	public static KeyEvent input(){
		if(MessagePanel.active!=null) MessagePanel.active.repaint();
		Interface.userinterface.getinput();
		return Interface.userinterface.keyevent;
	}

	/**
	 * Main output function for {@link BattleScreen}s.
	 *
	 * @param message Text to be printed.
	 * @param t TODO remove
	 * @param See {@link Javelin.Delay}.
	 */
	public static void message(final String out,final Javelin.Delay d){
		MessagePanel.active.add(out);
		switch(d){
			case WAIT:
				try{
					redraw();
					Thread.sleep(Preferences.messagewait);
				}catch(final InterruptedException e){
					e.printStackTrace();
				}
				MessagePanel.active.clear();
				break;
			case BLOCK:
				MessagePanel.active.add("\n"+TextZone.BLACK+"-- ENTER --");
				Javelin.delayblock=true;
				break;
		}
	}

	/**
	 * Calls {@link #prompt(String)}p until the user presses one of the allowed
	 * {@link Character}s.
	 *
	 * @param allowed <code>null</code> should never be included.
	 * @return User input, guaranteed to be among the allowed ones.
	 */
	public static Character prompt(String prompt,Set<Character> allowed){
		Character input=null;
		while(input==null||!allowed.contains(input))
			input=prompt(prompt);
		return input;
	}

	public static String group(List<?> foes){
		CountingSet count=new CountingSet();
		count.casesensitive=true;
		boolean hasunknown=false;
		for(Object c:foes){
			String foe=c.toString();
			if(foe.equals("?")){
				if(hasunknown) continue;
				hasunknown=true;
			}
			count.add(foe);
		}
		return count.toString();
	}

	/** @return A "Capitalized" version of the input. */
	public static String capitalize(String s){
		return Character.toUpperCase(s.charAt(0))+s.substring(1).toLowerCase();
	}
}
